--- a/conf/upsstats.html.sample
+++ b/conf/upsstats.html.sample
@@ -180,11 +180,4 @@ Network UPS Tools upsstats
 </td></tr>
 </table>
 
-<hr /><div><small>
-<a href="http://jigsaw.w3.org/css-validator/check/referer"><img style="float:right" src="http://jigsaw.w3.org/css-validator/images/vcss" alt="Valid CSS!" height="31" width="88"/></a>
-<a href="http://validator.w3.org/check?uri=referer"><img style="float:right"
-src="http://www.w3.org/Icons/valid-html40"
-alt="Valid HTML 4.0 Transitional" height="31" width="88"></a>
-</small></div>
-
 </body></html>
--- a/conf/upsstats-single.html.sample
+++ b/conf/upsstats-single.html.sample
@@ -286,9 +286,6 @@
 </tr>
 
 </table>
-<a href="http://validator.w3.org/check?uri=referer"><img style="float:right"
-        src="http://www.w3.org/Icons/valid-html40"
-        alt="Valid HTML 4.0 Transitional" height="31" width="88"></a>
 
 </body>
 </html>
