--- nut-2.7.3/tools/nut-scanner/scan_xml_http.c	2016-02-02 12:38:09.000000000 +0100
+++ nut-2.7.3AQU/tools/nut-scanner/scan_xml_http.c	2016-02-02 12:47:07.836429592 +0100
@@ -260,9 +260,8 @@
         fprintf(stderr,"Error creating socket\n");
         return NULL;
     }
-
 #define MAX 3
-    for (i = 0; i != MAX || current_nut_dev == NULL; i++) {
+    for (i = 0; i != MAX && current_nut_dev == NULL; i++) {
 		/* Initialize socket */
 		sockAddress.sin_family = AF_INET;
         //sockAddress.sin_addr.s_addr = INADDR_BROADCAST;
@@ -322,8 +321,7 @@
 					NI_NUMERICHOST) != 0) {
 
 					fprintf(stderr,
-						"Error converting IP address \
-						: %d, #%d/%d\n",errno);
+						"Error converting IP address: %d\n",errno);
                     usleep(usec_timeout);
 					continue;
 				}
