--- nut-2.0.1/configure.in.lib64	2004-08-12 05:50:00.000000000 -0400
+++ nut-2.0.1/configure.in	2005-03-22 12:09:21.933903641 -0500
@@ -667,10 +667,35 @@ AC_ARG_ENABLE(pkg-config,
 	esac
 ],
 [       
-	if (test -d /usr/lib/pkgconfig)
+    case $host in
+    *-*-linux*)
+      # Test if the compiler is 64bit
+      echo 'int i;' > conftest.$ac_ext
+      vorbis_cv_cc_64bit_output=no
+      if AC_TRY_EVAL(ac_compile); then
+        case `/usr/bin/file conftest.$ac_objext` in
+        *"ELF 64"*)
+          vorbis_cv_cc_64bit_output=yes
+          ;;
+        esac
+      fi
+      rm -rf conftest*
+      ;;
+    esac
+    
+    case $host_cpu:$vorbis_cv_cc_64bit_output in
+    powerpc64:yes | s390x:yes | sparc64:yes | x86_64:yes)
+      LIB="lib64"
+      ;;
+    *:*)
+      LIB="lib"
+      ;;
+    esac
+
+	if (test -d /usr/$LIB/pkgconfig)
 	then
 		AC_MSG_RESULT(yes)
-		PKG_CFG_DIR="/usr/lib/pkgconfig"
+		PKG_CFG_DIR="/usr/$LIB/pkgconfig"
 		USE_PKG_CFG=yes
 	else
 		AC_MSG_RESULT(no)
